File: core/src/ru/myitschool/cubegame/encounters/Encounter.java
package ru.myitschool.cubegame.encounters;

import com.badlogic.gdx.utils.Array;
import ru.myitschool.cubegame.entities.Entity;
import ru.myitschool.cubegame.utils.AdvancedArray;

/**
 * Created by Voyager on 17.02.2018.
 */
public abstract class Encounter implements Cloneable {

    public static final float FULL_TIME = 3f;
    private static AdvancedArray<Encounter> encounters = new AdvancedArray<Encounter>();
    private static Array<Encounter> nowPlayings = new Array<Encounter>();
    private Entity entity;
    private String text = "";
    private String name = "";
    private float time = 0f;

    public Encounter() {
        encounters.add(this);
    }

    public void trigger(Entity entity){
        addNowPlaying(this);
        this.entity = entity;
    }

    protected abstract void activate(Entity entity);

    public static void update(float delta) {
        for (Encounter encounter : nowPlayings) {
            if (encounter != null) {
                encounter.time += delta;
                if (encounter.time >= FULL_TIME) {
                    encounter.activate(encounter.entity);
                    encounter.time = 0f;
                    nowPlayings.removeValue(encounter, true);
                    encounter.entity = null;
                    encounter = null;
                }
            }
        }
    }
    public static void initEncouters(){
        new DartEncounter();
    }

    public static Encounter getEncounter(int index){
        return encounters.get(index).clone();
    }

    public static Encounter getRandomEncounter(){
        return encounters.getRandom().clone();
    }

    public String getText() {
        return text;
    }

    public String getName() {
        return name;
    }

    public void setText(String text) {
        this.text = text;
    }

    public void setName(String name) {
        this.name = name;
    }

    public static Array<Encounter> getNowPlayings() {
        return nowPlayings;
    }

    public static Encounter getNowPlaying(int index){
        return nowPlayings.get(index);
    }

    public static void addNowPlaying(Encounter nowPlaying) {
        nowPlayings.add(nowPlaying);
    }

    public static boolean hasNowPlayimg(){
        return nowPlayings.size > 0;
    }

    @Override
    protected Encounter clone(){
        try {
            Encounter encounter = (Encounter) super.clone();
            return encounter;
        } catch (Exception e){
            e.printStackTrace();
        }
        return null;
    }
}

File: core/src/com/wolg_vlad/dcrawler/encounters/Encounter.kt
package com.wolg_vlad.dcrawler.encounters

import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.event.EntityEvent
import com.wolg_vlad.dcrawler.event.EventController
import com.wolg_vlad.dcrawler.utils.AdvancedArray
import java.util.*

/**
 * Created by Voyager on 17.02.2018.
 */
abstract class Encounter : Cloneable {
    protected var entity: Entity? = null
    var text = ""
    var name = ""
    private var time = 0f
    fun trigger(entity: Entity) {
        addNowPlaying(this)
        this.entity = entity
        val map: HashMap<String, Any> = HashMap()
        map[EntityEvent.ENTITY_EXECUTOR_ARG_KEY] = entity
        map[EntityEvent.ENCOUNTER_ARG_KEY] = this
        EventController.callEvent(EntityEvent.ON_ENCOUNTER_EVENT, map)
    }

    protected abstract fun activate()

    override fun clone(): Encounter {
        return super.clone() as Encounter
    }

    companion object {
        const val FULL_TIME = 3f
        private val encounters = AdvancedArray<Encounter?>()
        var nowPlayings = mutableListOf<Encounter>()
        fun update(delta: Float) {
            nowPlayings = nowPlayings.filter { encounter ->
                encounter.time += delta
                if (encounter.time >= FULL_TIME) {
                    encounter.activate()
                    return@filter false
                }
                return@filter true
            }.toMutableList()
        }

        fun initEncouters() {
            DartEncounter()
        }

        fun getEncounter(index: Int): Encounter {
            return encounters[index]!!.clone()
        }

        val randomEncounter: Encounter
            get() = (encounters.random as Encounter).clone()

        fun getNowPlaying(index: Int): Encounter {
            return nowPlayings[index]
        }

        fun addNowPlaying(encounter: Encounter) {
            nowPlayings.add(encounter)
        }

        fun hasNowPlayimg(): Boolean {
            return nowPlayings.size > 0
        }
    }

    init {
        encounters.add(this)
    }
}